### PR TITLE
Introduce `jenkins.model.queue.QueueItem` interface

### PR DESCRIPTION
**Pull request by Vlatombe**
This formalizes methods used by `<t:queue>` taglib to do the rendering of queue items.

This will allow reuse of the taglib with alternative implementations of `QueueItem`.

Introduced in beta state to let it bake for some time until the interface is frozen.

### Proposed changelog entries

- Developer: Queue items elements are now formalized using `jenkins.model.queue.QueueItem`.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7926"></a>

### CODE REVIEW

File: core/src/main/java/hudson/widgets/BuildHistoryWidget.java
Comment (jglick) [inline core/src/main/java/hudson/widgets/BuildHistoryWidget.java:55]: Since we are incompatibly changing the signature, it would be a good time to restrict `BuildHistoryWidget`. (After verifying from https://github.com/jenkins-infra/usage-in-plugins that it is in fact not used from plugins. Unfortunately the default report https://ci.jenkins.io/job/Infra/job/usage-in-plugins/job/master/lastSuccessfulBuild/artifact/output/ shows only usages of _deprecated_ members so it is useless for this purpose.)

File: core/src/main/java/jenkins/model/queue/QueueItem.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/QueueItem.java:13]:
```suggestion
 * Interface used by Jelly views to render a queue item through {@code <t:queue>}.
```

File: core/src/main/java/jenkins/model/queue/QueueItem.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/QueueItem.java:76]: I think you meant

```suggestion
        return getTask().getFullDisplayName();
```

?

File: core/src/main/java/jenkins/model/queue/QueueItem.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/QueueItem.java:71]:
```suggestion
     * @return a display name for this queue item; by default, {@link Queue.Task#getFullDisplayName()}
```

File: core/src/test/java/hudson/model/MockItem.java
Comment (jglick) [inline core/src/test/java/hudson/model/MockItem.java:95]: BTW these two could just get `default` implementations in `Task`. Normally the impl class is picking up defaults from something else like `Job`, but it would be useful here.

**Review by jglick — CHANGES_REQUESTED** [review]
`displayName` vs. `fullDisplayName` is probably a regression. Otherwise looks good to me.

**Review by jglick — APPROVED** [review]

File: core/src/main/java/hudson/widgets/BuildHistoryWidget.java
Comment (Vlatombe) [reply core/src/main/java/hudson/widgets/BuildHistoryWidget.java:55]: https://github.com/search?q=org%3Ajenkinsci+BuildHistoryWidget&type=code shows usage in
* batch-task-plugin -- https://github.com/jenkinsci/batch-task-plugin/blob/1b92e1749646186e73e076f9a9b41e48ba6f8455/src/main/java/hudson/plugins/batch_task/BatchTask.java#L20
* DotCI -- https://github.com/jenkinsci/DotCi/blob/611dae322da72898ed75bcf635da89cb24b1cbd7/src/main/java/com/groupon/jenkins/dynamic/build/DynamicSubProject.java#L48. This extends `Project` so the method should be deleted.

File: core/src/main/java/hudson/model/Queue.java
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java:1952]: Should defaults here check `this instanceof AccessControlled`? Again unlikely to matter in production code which should pick up e.g. https://github.com/jenkinsci/jenkins/blob/6a1b865270c2a1584d79928adf01ab78af8cd63d/core/src/main/java/hudson/model/AbstractProject.java#L1175-L1183

:thinking: check `QueueTaskFilter`

**Review by NotMyFault — APPROVED** [review]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

File: core/src/main/java/hudson/widgets/BuildHistoryWidget.java
Comment (lacostej) [reply core/src/main/java/hudson/widgets/BuildHistoryWidget.java:55]: The BuildHistoryWidget feature is broken in the `batch-task-plugin` and I wish to get it to work again. [I haven't been able to use the Job/Run implementations](https://issues.jenkins.io/browse/JENKINS-33833), and wonder if I can get it to work with the QueueItem one. Is there any guideline on how to use @Beta features?

Comment (jglick) [reply core/src/main/java/hudson/widgets/BuildHistoryWidget.java:55]: > how to use beta features?

```xml
<useBeta>true</useBeta>
```

in your POM.

**Pull request merged**
